src/pudl/helpers.py

```diff
@@ -1441,6 +1446,67 @@ def dedupe_on_category(
     return dedup_df.drop_duplicates(subset=base_cols, keep="first")
 
 
+def dedupe_and_drop_nas(
+    dedup_df: pd.DataFrame,
+    primary_key_cols: list[str],
+) -> pd.DataFrame:
+    """Deduplicate a df by comparing primary key columns and dropping null rows.
+
+    When a primary key appears twice in a dataframe, and one record is all null other
+    than the primary keys, drop the null row.
+
+    Args:
+        dedup_df: the dataframe with the records to deduplicate.
+        primary_key_cols: list of columns which must not be duplicated.
+
+    Returns:
+        The deduplicated dataframe.
+    """
+    dupes = dedup_df.loc[dedup_df.duplicated(subset=primary_key_cols, keep=False)]
+    dupe_groups = dupes.groupby(primary_key_cols)
+    if (dupe_groups.nunique() > 1).any().any():  # noqa: PD101
+        raise AssertionError(
+            f"Duplicate records with disagreeing data: {dupes[dupes.set_index(primary_key_cols).index.duplicated(keep=False)]}"
+        )
+    deduped = dupe_groups.first().reset_index()
+    # replace the duplicated rows with the deduped versions
+    return pd.concat(
+        [dedup_df.drop_duplicates(subset=primary_key_cols, keep=False), deduped],
+        ignore_index=True,
+    )
+
+
+def standardize_percentages_ratio(
+    frac_df: pd.DataFrame,
+    mixed_cols: list[str],
+) -> pd.DataFrame:
+    """Standardize mixed 0-1 and 0-100 percentage/ratio reporting in a column.
+
+    When a column uses both 0-1 and 0-100 scales to describe percentages, standardize
+    and convert to 0-1 ratios/fractions.
+
+    Args:
+        frac_df: the dataframe with the columns to standardize.
+        mixed_cols: list of columns which should get standardized to the 0-1 scale.
+
+    Returns:
+        The standardized dataframe.
+    """
+    logger.info(f"Standardizing ratios and percentages for {mixed_cols}")
+    for col in mixed_cols:
+        if not pd.api.types.is_numeric_dtype(frac_df[col]):
+            raise AssertionError(
+                f"{col}: Standardization method requires numeric dtype."
+            )
+        frac_df.loc[(frac_df[col] > 1) & (frac_df[col] <= 100), col] /= 100
```

Comment:
Is there any way we can guard against the case where the reported value is between 0 and 1, and it's supposed to represent a fraction of a percent? This is a modest issue if the values overall are evenly distributed across the whole range from 0 to 100%, but if the expected values are on the order of 1% it could be quite wrong.

In some EIA data, we've found that the reporting format varies by year, with some years using 0.0-1.0 and others using 0-100%, and have restricted this kind of reformatting to apply only to the years reporting percentages.  In years with mixed reporting formats, it might also be the case that within a given utility's data, the format is consistent.